swift/extractor/infra/log/SwiftLogging.h

```diff
@@ -0,0 +1,164 @@
+
+// only do the actual logging if the picked up `Logger` instance is configured to handle the
+// provided log level
+#define LOG_IMPL(severity, ...)                                                        \
+  do {                                                                                 \
+    if (auto& _logger = logger(); severity >= _logger.level()) {                       \
+      BINLOG_CREATE_SOURCE_AND_EVENT(_logger.writer(), severity, , binlog::clockNow(), \
+                                     __VA_ARGS__);                                     \
+    }                                                                                  \
+  } while (false)
+
+namespace codeql {
+
+// This class is responsible for the global log state (outputs, log level rules, flushing)
+// State is stored in the singleton `Log::instance()`.
+// Before using logging, `Log::configure("<name>")` should be used (e.g.
+// `Log::configure("extractor")`). Then, `Log::flush()` should be regularly called.
+class Log {
+ public:
+  using Level = binlog::Severity;
+
+ private:
+  // Output filtered according to a configured log level
+  template <typename Output>
+  struct FilteredOutput {
+    binlog::Severity level;
+    Output output;
+    binlog::EventFilter filter{
+        [this](const binlog::EventSource& src) { return src.severity >= level; }};
+
+    template <typename... Args>
+    FilteredOutput(Level level, Args&&... args)
+        : level{level}, output{std::forward<Args>(args)...} {}
+
+    FilteredOutput& write(const char* buffer, std::streamsize size) {
+      filter.writeAllowed(buffer, size, output);
+      return *this;
+    }
+
+    // if configured as `no_logs`, the output is effectively disabled
+    explicit operator bool() const { return level < Level::no_logs; }
+  };
+
+  using LevelRule = std::pair<std::regex, Level>;
+  using LevelRules = std::vector<LevelRule>;
+
+  static constexpr const char* format = "%u %S [%n] %m (%G:%L)\n";
+
+  binlog::Session session;
+  std::ofstream textFile;
+  FilteredOutput<std::ofstream> binary{Level::no_logs};
+  FilteredOutput<binlog::TextOutputStream> text{Level::info, textFile, format};
+  FilteredOutput<binlog::TextOutputStream> console{Level::warning, std::cerr, format};
+  LevelRules sourceRules;
+  std::string rootName;
+
+  Log() = default;
+
+  static Log& instance() {
+    static Log ret;
+    return ret;
+  }
+
+  friend class Logger;
+  friend binlog::Session;
+
+  Level getLevelForSource(std::string_view name) const;
+  Log& write(const char* buffer, std::streamsize size);
+  static class Logger& logger();
+
+ public:
+  // Configure logging. This consists in
+  // * setting up a default logger with `root` as name
+  // * using environment variable `CODEQL_EXTRACTOR_SWIFT_LOG_DIR` to choose where to dump the log
+  //   file(s). Log files will go to a subdirectory thereof named after `root`
+  // * using environment variable `CODEQL_EXTRACTOR_SWIFT_LOG_LEVELS` to configure levels for
```

Comment:
Thinking about different types of doc reader:
 1. Extractor developers. They're probably the main target audience, at least at this point, for debugging the extractor. Do we already have any docs aimed at developers besides the readme? For C++, we have this sort of thing on the extractor team wiki.
 2. Field teams. I'm not sure what sort of support offering we plan to have for Swift (or what we're doing for the public beta), but if anyone is going to be working with users to debug database creation, they'll probably need to tweak logging too. We'd probably want those docs in a wiki, possibly the same one as for extractor developers, but should point the field teams at it.
 3. External users. If we want them to see it, that should probably go through the doc team. But my hope is that they shouldn't need to tweak these options at all, unless instructed by one of us on an issue (in which case we can tell them exactly what string would be useful and how to set it).

As a first step, let's aim it at extractor developers since we *definitely* need it. For discoverability, I might suggest creating a `codeql/swift/extractor/README.md` or adding it to `codeql/swift/README.md`, which already has a section on debugging. For field teams/external users, we can either pro-actively ask the field and docs teams whether they'd like this documented and how, or we can wait for a request.

Reply:
Added a section to `swift/README.md`. I will ask how to circulate this info to the field team. I agree this should probably not be visible up front in user documentation.